rules/REQUEST-922-MULTIPART-ATTACK.conf

```diff
@@ -0,0 +1,116 @@
+# ------------------------------------------------------------------------
+# OWASP ModSecurity Core Rule Set ver.4.0.0-rc1
+# Copyright (c) 2006-2020 Trustwave and contributors. All rights reserved.
+# Copyright (c) 2021-2022 Core Rule Set project. All rights reserved.
+#
+# The OWASP ModSecurity Core Rule Set is distributed under
+# Apache Software License (ASL) version 2
+# Please see the enclosed LICENSE file for full details.
+# ------------------------------------------------------------------------
+
+#
+# -= Paranoia Level 0 (empty) =- (apply unconditionally)
+#
+
+# This file is to address the 3UWMWA6W vulnerability.
+# It requires the latest version of ModSecurity 2.x/3.x or newer, where the variable
+# MULTIPART_PART_HEADERS was introduced.
+
+SecRule TX:DETECTION_PARANOIA_LEVEL "@lt 1" "id:922011,phase:1,pass,nolog,skipAfter:END-REQUEST-922-MULTIPART-ENFORCEMENT"
+SecRule TX:DETECTION_PARANOIA_LEVEL "@lt 1" "id:922012,phase:2,pass,nolog,skipAfter:END-REQUEST-922-MULTIPART-ENFORCEMENT"
+
+#
+# -= Paranoia Level 1 (default) =- (apply only when tx.detection_paranoia_level is sufficiently high: 1 or higher)
+#
+
+# Only allow specific charsets when using "_charset_"
+# Note: this is in phase:2 because these are headers that come in the body
+SecRule &MULTIPART_PART_HEADERS:_charset_ "!@eq 0" \
```

Comment:
Does this also need to be a `@rx` like below? Why isn't the `Content-Transfer-Encoding` header matched with `&` and `@eq`?

Reply:
This is the difficult part to understand from the new MULTIPART_PART_HEADERS collection. 

The key in the collection is the `name` part of the `Content-Disposition` multipart header. So, for `Content-disposition: form-data; name="_charset_"`, you will find this in the correct place: `MULTIPART_PART_HEADERS:_charset_`.

But for `Content-Transfer-Encoding`, it is part of the headers that are keyed by a different name, the name of the `Content-disposition` part on where they are.

In the example tests there is this:
```
Content-Disposition: form-data; name="flavors"
Content-Transfer-Encoding: 8bit
Content-Type: text/plain; charset=utf-7
```

The `Content-Transfer-Encoding` here will be indexed the `MULTIPART_PART_HEADERS:flavors` variable. That's why we need to look for regular expressions in all keys, and we cannot use a particular index for that. Consequently, we cannot match with & because that might happen in any key in the MULTIPART_PART_HEADERS collection.